{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c39f9cbc",
   "metadata": {},
   "source": [
    "### This notebook will contain the loading of the datasets, and tie together the different parts of the project"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae572cff",
   "metadata": {},
   "source": [
    "### IMPORTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0048a912",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from PriceDataLoader import PriceDataLoader\n",
    "from WeatherDataLoader import WeatherDataLoader\n",
    "from models.CNN import *\n",
    "from models.FFN import *\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "from models.CointegrationResidualGenerator import CointegrationResidualGenerator\n",
    "from models.BacktestSharpeEvaluator import BacktestSharpeEvaluator\n",
    "from PortfolioOptimizer import PortfolioOptimizer\n",
    "from TimeSeriesDataSplitter import TimeSeriesDataSplitter\n",
    "from Utility import Utility\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86c966b7",
   "metadata": {},
   "source": [
    "### Data Preperation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d7fdc4ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data loaded successfully from ../data/european_wholesale_electricity_price_data_daily.csv\n"
     ]
    }
   ],
   "source": [
    "# Initialize the DataLoader with the test data\n",
    "price_parser = PriceDataLoader(file_path=\"../data/european_wholesale_electricity_price_data_daily.csv\")\n",
    "\n",
    "# Get list of countries\n",
    "all_countries_list = price_parser.get_country_list()\n",
    "small_counties_list = all_countries_list[:5]\n",
    "\n",
    "# Define the total time range\n",
    "train_time_range = \"2015-01-01,2024-01-01\"\n",
    "\n",
    "# Define the window size and stride\n",
    "one_window_days = 365\n",
    "window_stride_days = 5\n",
    "\n",
    "# Get the rolling window of the test data\n",
    "x_data = price_parser.get_price_matrix_rolling_window(\n",
    "    one_window_days=one_window_days,\n",
    "    window_stride_days=window_stride_days,\n",
    "    time_range=train_time_range,\n",
    "    countries=all_countries_list,\n",
    "    fill_method=\"ffill\",\n",
    ")\n",
    "\n",
    "# -----------------------------------------------------\n",
    "\n",
    "# Get the price matrix for the training/testing/validation data\n",
    "price_matrix = price_parser.get_price_matrix(\n",
    "    time_range=train_time_range,\n",
    "    countries=all_countries_list,\n",
    "    fill_method=\"ffill\",\n",
    ")\n",
    "\n",
    "# ------------------------------------------------------\n",
    "# Get the returns matrix\n",
    "returns_tr = price_matrix.pct_change().dropna()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "acb2ee39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "Skipping due to error: Found array with 0 sample(s) (shape=(0, 30)) while a minimum of 1 is required by LinearRegression.\n",
      "The cumulative residual window size exceeds the available data.\n",
      "The cumulative residual window size exceeds the available data.\n",
      "The cumulative residual window size exceeds the available data.\n",
      "The cumulative residual window size exceeds the available data.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The cumulative residual window size exceeds the available data.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n",
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "# Get the rolling window of the training data\n",
    "x_cumulative_residuals, y_next_day_returns = Utility.process_training_data(\n",
    "    x_tr_data=x_data,\n",
    "    returns=returns_tr,\n",
    "    cumulative_residual_window=30\n",
    ")\n",
    "\n",
    "# Convert x_tr_cumulative_residuals to a numpy array\n",
    "x_cumulative_residuals = np.array(x_cumulative_residuals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c011bd2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------------------\n",
      "x_cumulative_residuals shape: (111779, 31, 30)\n",
      "x_cumulative_residuals type: <class 'numpy.ndarray'>\n",
      "y_next_day_returns shape: (111779, 31)\n",
      "y_next_day_returns type: <class 'pandas.core.frame.DataFrame'>\n"
     ]
    }
   ],
   "source": [
    "# # Print the shape of the cnn_input_array\n",
    "# print(f\"cnn_input_array shape: {cnn_input_array.shape}\")\n",
    "# # Print the type of the cnn_input_array\n",
    "# print(f\"cnn_input_array type: {type(cnn_input_array)}\")\n",
    "\n",
    "# # Print the shape of next_day_returns\n",
    "# print(f\"next_day_returns shape: {next_day_returns.shape}\")\n",
    "# # Print the type of next_day_returns\n",
    "# print(f\"next_day_returns type: {type(next_day_returns)}\")\n",
    "\n",
    "print(\"------------------------------------\")\n",
    "\n",
    "# Print the shape of a3_0\n",
    "print(f\"x_cumulative_residuals shape: {x_cumulative_residuals.shape}\")\n",
    "# Print the type of a3_0\n",
    "print(f\"x_cumulative_residuals type: {type(x_cumulative_residuals)}\")\n",
    "\n",
    "# Print the shape of a3_1\n",
    "print(f\"y_next_day_returns shape: {y_next_day_returns.shape}\")\n",
    "# Print the type of a3_1\n",
    "print(f\"y_next_day_returns type: {type(y_next_day_returns)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "382150bc",
   "metadata": {},
   "source": [
    "### CNN + FFN (Ablation study)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c729c902",
   "metadata": {},
   "source": [
    "#### Sharpe ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8397151f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sharpe_ratio_loss(returns, risk_free_rate=0.0):\n",
    "    \"\"\"\n",
    "    Custom loss function to maximize the Sharpe Ratio.\n",
    "    Args:\n",
    "        returns (torch.Tensor): Predicted returns\n",
    "        risk_free_rate (float): Risk-free rate for Sharpe calculation\n",
    "    Returns:\n",
    "        torch.Tensor: Negative Sharpe Ratio (to minimize)\n",
    "    \"\"\"\n",
    "    excess_returns = returns - risk_free_rate\n",
    "    mean_excess = torch.mean(excess_returns)\n",
    "    std_excess = torch.std(excess_returns, unbiased=False) + 1e-6  # epsilon for stability\n",
    "    sharpe_ratio = mean_excess / std_excess\n",
    "    return -sharpe_ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5ab69198",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data loaded successfully from ../data/european_wholesale_electricity_price_data_daily.csv\n",
      "Weather data loaded successfully from ../data/aggregated_weather.csv\n",
      "Number of countries: 31\n",
      "Number of features: 3\n",
      "Weather rolling window shape: 336 x 30\n",
      "(329, 89, 30)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "### -------------- PARSING DATA -------------- ###\n",
    "price_parser = PriceDataLoader(file_path=\"../data/european_wholesale_electricity_price_data_daily.csv\")\n",
    "weather_parser = WeatherDataLoader(file_path=\"../data/aggregated_weather.csv\")\n",
    "\n",
    "# Get list of countries\n",
    "all_countries_list = price_parser.get_country_list()\n",
    "print(f\"Number of countries: {len(all_countries_list)}\")\n",
    "\n",
    "# Get daily price matrix for all countries for 2021\n",
    "price_matrix = price_parser.get_price_matrix(\n",
    "    time_range=\"2021-01-01,2021-12-31\",\n",
    "    countries=all_countries_list,\n",
    "    fill_method=\"ffill\"\n",
    ")\n",
    "\n",
    "# Get daily returns\n",
    "returns = price_matrix.pct_change().dropna()\n",
    "\n",
    "# Get weather matrix with selected features\n",
    "weather_features = ['temperature_2m_mean', 'wind_speed_mean']  # Adjust based on available features\n",
    "\n",
    "print(f\"Number of features: {len(weather_features) + 1}\") # +1 due to price\n",
    "\n",
    "weather_matrix = weather_parser.get_weather_matrix(\n",
    "    time_range=\"2021-01-01,2021-12-31\",\n",
    "    countries=all_countries_list,\n",
    "    fill_method=\"ffill\",\n",
    "    features=weather_features\n",
    ")\n",
    "\n",
    "### -------------- COINTEGRATION RESIDUALS -------------- ###\n",
    "residual_generator = CointegrationResidualGenerator(price_matrix)\n",
    "residual_generator.compute_all_asset_residuals()\n",
    "asset_residuals = residual_generator.get_asset_residuals()\n",
    "\n",
    "# Prepare CNN input from residuals\n",
    "cumulative_residual_window = 30\n",
    "cnn_residual_input = residual_generator.prepare_cnn_input_from_residuals(window=cumulative_residual_window)\n",
    "\n",
    "# Prepare weather input with rolling windows\n",
    "weather_rolling_windows = weather_parser.get_weather_matrix_rolling_window(\n",
    "    one_window_days=cumulative_residual_window,\n",
    "    window_stride_days=1,\n",
    "    time_range=\"2021-01-01,2021-12-31\",\n",
    "    countries=all_countries_list,\n",
    "    fill_method=\"ffill\",\n",
    "    features=weather_features\n",
    ")\n",
    "print(f\"Weather rolling window shape: {len(weather_rolling_windows)} x {len(weather_rolling_windows[0])}\")\n",
    "\n",
    "# Ensure alignment between residual and weather windows\n",
    "# cnn_residual_input shape: [samples, window, assets]\n",
    "# weather_rolling_windows: List of DataFrames, each [window, (feature, country)]\n",
    "num_samples = len(cnn_residual_input)\n",
    "combined_input = []\n",
    "for i in range(num_samples):\n",
    "    if i < len(weather_rolling_windows):\n",
    "        weather_window = weather_rolling_windows[i]\n",
    "        # Flatten MultiIndex columns to [country_feature]\n",
    "        weather_array = weather_window.values  # Shape: [window, num_features * num_countries]\n",
    "        residual_array = cnn_residual_input[i]  # Shape: [window, num_countries]\n",
    "        # Concatenate residuals and weather features along feature dimension\n",
    "        combined_window = np.concatenate([residual_array, weather_array], axis=1)  # Shape: [window, num_countries + num_features * num_countries]\n",
    "        combined_input.append(combined_window)\n",
    "combined_input_array = np.array(combined_input)  # Shape: [samples, window, num_countries + num_features * num_countries]\n",
    "\n",
    "# Get next-day returns\n",
    "start_idx_in_returns = returns.index.get_loc(asset_residuals.index[0])\n",
    "next_day_indices = [start_idx_in_returns + i + cumulative_residual_window for i in range(num_samples)]\n",
    "next_day_returns = returns.iloc[next_day_indices]\n",
    "\n",
    "# Transform combined input for CNN\n",
    "cnn_input_array = combined_input_array.transpose(0, 2, 1)  # [samples, features, window]\n",
    "\n",
    "print(cnn_input_array.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0935612c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### -------------- FEED THE 30-DAY CUMULATIVE RESIDUALS OF EVERY COUNTRY TO CNN+FFN -------------- ###\n",
    "# Transform cnn_input to be compatible with the CNN input shape\n",
    "#cnn_input_array = cnn_input.transpose(0, 2, 1) # [samples, features, window]\n",
    "\n",
    "# FILIP'S SECTION: CNN+FFN\n",
    "# Hey Filip, this is the section where you can add your code to train the CNN+FFN model.\n",
    "# cnn_input_array essentially contains 329 training data points, each data point is 30-day cumulative residuals for the 31 countries.\n",
    "# So, for one \"data point\" you would feed the set of 30-day cumulative residuals for every country to the CNN+FFN model.\n",
    "# One data point should result in a set of weights for each of the 31 countries.\n",
    "# Each set of weights is used to calculate one next-day portfolio return.\n",
    "# Repeat this for all 329 data points to get a set of 329 portfolio returns.\n",
    "# You can then use these portfolio returns to calculate the Sharpe ratio\n",
    "# Optimize the CNN+FFN to maximize the Sharpe ratio.\n",
    "\n",
    "# Set the device to GPU if available, otherwise MPS (for Mac silicon) or CPU\n",
    "if torch.cuda.is_available():\n",
    "    device = torch.device(\"cuda\")\n",
    "elif torch.backends.mps.is_available():\n",
    "    device = torch.device(\"mps\")\n",
    "else:\n",
    "    device = torch.device(\"cpu\")\n",
    "\n",
    "torch.manual_seed(1)  # For reproducibility\n",
    "\n",
    "# Create the PortfolioOptimizer instance and train the model\n",
    "# NOTE: We need to have some validation data to evaluate the model performance\n",
    "# otherwise we will be overfitting the model to the training data\n",
    "\n",
    "# Grid search for hyperparameters\n",
    "num_filters = [8, 16, 32]\n",
    "filter_sizes = [3, 5, 7]\n",
    "hidden_dims = [64, 128]\n",
    "\n",
    "# Choose the best hyperparameters based on validation performance\n",
    "best_score = float('-inf')\n",
    "best_params = None\n",
    "\n",
    "# Training and validation split\n",
    "# 70% training, 15% validation, 15% testing\n",
    "# Note: The test set is not used in the training process, but it can be used ONCE to evaluate the final model performance\n",
    "# The validation set is used to tune the hyperparameters and avoid overfitting\n",
    "\"\"\"\n",
    "train_dataset = cnn_input_array[:int(0.7 * len(cnn_input_array))]\n",
    "train_next_day_returns = next_day_returns[:int(0.7 * len(next_day_returns))]\n",
    "\n",
    "val_dataset = cnn_input_array[int(0.7 * len(cnn_input_array)):]\n",
    "val_next_day_returns = next_day_returns[int(0.7 * len(next_day_returns)):]\n",
    "\n",
    "test_dataset = val_dataset[int(0.5 * len(val_dataset)):]\n",
    "test_next_day_returns = val_next_day_returns[int(0.5 * len(val_next_day_returns)):]\n",
    "\"\"\"\n",
    "# datasetSplitter = TimeSeriesDataSplitter(data_x=cnn_input_array, data_y=next_day_returns, train_size=0.7, validation_size=0.15, test_size=0.15)\n",
    "datasetSplitter = TimeSeriesDataSplitter(data_x=x_cumulative_residuals, data_y=y_next_day_returns, train_size=0.7, validation_size=0.15, test_size=0.15)\n",
    "(train), (val), (test) = datasetSplitter.split_data()\n",
    "\n",
    "# Unpack the datasets\n",
    "train_dataset, train_next_day_returns = train\n",
    "val_dataset, val_next_day_returns = val\n",
    "test_dataset, test_next_day_returns = test\n",
    "\n",
    "print(f\"Train dataset shape: {train_dataset.shape}, Train next day returns shape: {train_next_day_returns.shape}\")\n",
    "print(f\"Validation dataset shape: {val_dataset.shape}, Validation next day returns shape: {val_next_day_returns.shape}\")\n",
    "print(f\"Test dataset shape: {test_dataset.shape}, Test next day returns shape: {test_next_day_returns.shape}\")\n",
    "\n",
    "for num_filter in num_filters:\n",
    "    for filter_size in filter_sizes:\n",
    "        for hidden_dim in hidden_dims:\n",
    "                print(f\"Training with filters: {num_filter}, filter size: {filter_size}, hidden dim: {hidden_dim}\")\n",
    "                \n",
    "                # Initialize the PortfolioOptimizer\n",
    "                optimizer = PortfolioOptimizer(train_dataset, \n",
    "                                               train_next_day_returns, \n",
    "                                               val_dataset,\n",
    "                                               val_next_day_returns,\n",
    "                                               batch_size=1000, \n",
    "                                               num_epochs=1000, \n",
    "                                               num_filters=num_filter, \n",
    "                                               device=device, \n",
    "                                               filter_size=filter_size,\n",
    "                                               hidden_dim=hidden_dim,\n",
    "                                               patience=20,\n",
    "                                               use_transformer=True)\n",
    "                \n",
    "                # Train the model\n",
    "                _, portfolio_returns = optimizer.train(verbose=True) # The final sharpe should be from the test set\n",
    "\n",
    "                # Evaluate the model's hyperparameters on the validation set\n",
    "                val_sharpe, _ = optimizer.evaluate_final()\n",
    "                if val_sharpe > best_score:\n",
    "                    best_score = val_sharpe\n",
    "                    best_params = (num_filter, filter_size, hidden_dim)\n",
    "                    best_portfolio_returns = portfolio_returns\n",
    "    \n",
    "print(f\"Best parameters: {best_params} with Sharpe Ratio on Validation Set: {best_score}\")\n",
    "\n",
    "### -------------- GET THE WEIGHTS OUTPUTTED FROM CNN+FFN -------------- ###\n",
    "# Initializing the Sharpe ratio evaluator\n",
    "#evaluator = BacktestSharpeEvaluator()\n",
    "\n",
    "# Get the weight outputted from the CNN+FFN model\n",
    "# One weight for each country\n",
    "#weights = np.array([]) # CHANGE THIS TO THE ACTUAL WEIGHTS OUTPUTTED FROM THE CNN+FFN MODEL\n",
    "\n",
    "# Normalize the weights using L1 normalization\n",
    "# This is done to ensure that the portfolio is dollar-neutral\n",
    "#normalized_weights = evaluator.normalize_weights_l1(weights)\n",
    "\n",
    "# Multiply the normalized weights (vector) with the next-day returns (vector) to get the portfolio return\n",
    "#next_day_portfolio_return = evaluator.compute_portfolio_return(normalized_weights, next_day_returns.iloc[0].values)\n",
    "\n",
    "# Store the next day portfolio return \n",
    "#evaluator.add_return(next_day_portfolio_return)\n",
    "\n",
    "# Repeat the above step for all 329 data points, adding the portfolio returns to the evaluator for each data point\n",
    "\n",
    "# Once all portfolio returns are calculated, you can calculate the Sharpe ratio\n",
    "\n",
    "# Train the model to optimize the Sharpe Ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90dd3fc4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using device: mps\n",
      "Data loaded successfully from ../data/european_wholesale_electricity_price_data_daily.csv\n",
      "Weather data loaded successfully from ../data/aggregated_weather.csv\n",
      "Number of countries: 31\n",
      "Number of total features: 10\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/filip/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:393: RuntimeWarning: invalid value encountered in log\n",
      "  result = func(self.values, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2181, 31)\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[10]\u001b[39m\u001b[32m, line 45\u001b[39m\n\u001b[32m     43\u001b[39m window_size = \u001b[32m30\u001b[39m\n\u001b[32m     44\u001b[39m stride = \u001b[32m1\u001b[39m\n\u001b[32m---> \u001b[39m\u001b[32m45\u001b[39m combined_data = \u001b[43mdata_prep\u001b[49m\u001b[43m.\u001b[49m\u001b[43mprepare_rolling_windows\u001b[49m\u001b[43m(\u001b[49m\u001b[43mwindow_size\u001b[49m\u001b[43m=\u001b[49m\u001b[43mwindow_size\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstride\u001b[49m\u001b[43m=\u001b[49m\u001b[43mstride\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     46\u001b[39m next_day_returns = data_prep.prepare_next_day_returns(returns=returns, window_size=window_size, stride=stride)\n\u001b[32m     47\u001b[39m (train_data, train_returns), (val_data, val_returns), (test_data, test_returns) = \\\n\u001b[32m     48\u001b[39m     data_prep.create_train_val_test_split(combined_data=combined_data, next_day_returns=next_day_returns)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/stat7007/project/src/DataPreparation.py:34\u001b[39m, in \u001b[36mprepare_rolling_windows\u001b[39m\u001b[34m(self, window_size, stride)\u001b[39m\n\u001b[32m     28\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34malign_data_dates\u001b[39m(\u001b[38;5;28mself\u001b[39m) -> Tuple[pd.DataFrame, pd.DataFrame]:\n\u001b[32m     29\u001b[39m \u001b[38;5;250m    \u001b[39m\u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m     30\u001b[39m \u001b[33;03m    Align price residuals and weather data to have the same dates.\u001b[39;00m\n\u001b[32m     31\u001b[39m \n\u001b[32m     32\u001b[39m \u001b[33;03m    Returns:\u001b[39;00m\n\u001b[32m     33\u001b[39m \u001b[33;03m        Tuple[pd.DataFrame, pd.DataFrame]: Aligned price residuals and weather data DataFrames with common dates.\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m34\u001b[39m \u001b[33;03m    \"\"\"\u001b[39;00m\n\u001b[32m     35\u001b[39m     \u001b[38;5;66;03m# Find common dates between price residuals and weather data\u001b[39;00m\n\u001b[32m     36\u001b[39m     common_dates = \u001b[38;5;28mself\u001b[39m.price_residuals.index.intersection(\u001b[38;5;28mself\u001b[39m.weather_data.index)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/frame.py:4081\u001b[39m, in \u001b[36mDataFrame.__getitem__\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   4078\u001b[39m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m._get_item_cache(key)\n\u001b[32m   4080\u001b[39m     \u001b[38;5;28;01melif\u001b[39;00m is_mi \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m.columns.is_unique \u001b[38;5;129;01mand\u001b[39;00m key \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m.columns:\n\u001b[32m-> \u001b[39m\u001b[32m4081\u001b[39m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_getitem_multilevel\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   4083\u001b[39m \u001b[38;5;66;03m# Do we have a slicer (on rows)?\u001b[39;00m\n\u001b[32m   4084\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, \u001b[38;5;28mslice\u001b[39m):\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/frame.py:4189\u001b[39m, in \u001b[36mDataFrame._getitem_multilevel\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   4186\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n\u001b[32m   4187\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m   4188\u001b[39m     \u001b[38;5;66;03m# loc is neither a slice nor ndarray, so must be an int\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m4189\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_ixs\u001b[49m\u001b[43m(\u001b[49m\u001b[43mloc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m1\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/frame.py:4008\u001b[39m, in \u001b[36mDataFrame._ixs\u001b[39m\u001b[34m(self, i, axis)\u001b[39m\n\u001b[32m   4004\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n\u001b[32m   4006\u001b[39m \u001b[38;5;66;03m# icol\u001b[39;00m\n\u001b[32m   4007\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m-> \u001b[39m\u001b[32m4008\u001b[39m     label = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m[\u001b[49m\u001b[43mi\u001b[49m\u001b[43m]\u001b[49m\n\u001b[32m   4010\u001b[39m     col_mgr = \u001b[38;5;28mself\u001b[39m._mgr.iget(i)\n\u001b[32m   4011\u001b[39m     result = \u001b[38;5;28mself\u001b[39m._box_col_values(col_mgr, i)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/indexes/multi.py:2192\u001b[39m, in \u001b[36mMultiIndex.__getitem__\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   2190\u001b[39m             retval.append(np.nan)\n\u001b[32m   2191\u001b[39m         \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m-> \u001b[39m\u001b[32m2192\u001b[39m             retval.append(\u001b[43mlev\u001b[49m\u001b[43m[\u001b[49m\u001b[43mlevel_codes\u001b[49m\u001b[43m[\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m)\n\u001b[32m   2194\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mtuple\u001b[39m(retval)\n\u001b[32m   2195\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m   2196\u001b[39m     \u001b[38;5;66;03m# in general cannot be sure whether the result will be sorted\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/stat7007/project/venv/lib/python3.12/site-packages/pandas/core/indexes/base.py:5373\u001b[39m, in \u001b[36mIndex.__getitem__\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   5369\u001b[39m \u001b[38;5;129m@final\u001b[39m\n\u001b[32m   5370\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34m__setitem__\u001b[39m(\u001b[38;5;28mself\u001b[39m, key, value) -> \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m   5371\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[33m\"\u001b[39m\u001b[33mIndex does not support mutable operations\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m-> \u001b[39m\u001b[32m5373\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34m__getitem__\u001b[39m(\u001b[38;5;28mself\u001b[39m, key):\n\u001b[32m   5374\u001b[39m \u001b[38;5;250m    \u001b[39m\u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m   5375\u001b[39m \u001b[33;03m    Override numpy.ndarray's __getitem__ method to work as desired.\u001b[39;00m\n\u001b[32m   5376\u001b[39m \n\u001b[32m   (...)\u001b[39m\u001b[32m   5382\u001b[39m \n\u001b[32m   5383\u001b[39m \u001b[33;03m    \"\"\"\u001b[39;00m\n\u001b[32m   5384\u001b[39m     getitem = \u001b[38;5;28mself\u001b[39m._data.\u001b[34m__getitem__\u001b[39m\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: "
     ]
    }
   ],
   "source": [
    "from Trainer import Trainer\n",
    "from PortfolioOptimizer import PortfolioOptimizer\n",
    "from DataPreparation import DataPreparation\n",
    "\n",
    "# Set device\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else\n",
    "                        \"mps\" if torch.backends.mps.is_available() else \"cpu\")\n",
    "print(f\"Using device: {device}\")\n",
    "\n",
    "# 1. Load and prepare data\n",
    "price_parser = PriceDataLoader(file_path=\"../data/european_wholesale_electricity_price_data_daily.csv\")\n",
    "weather_parser = WeatherDataLoader(file_path=\"../data/aggregated_weather.csv\")\n",
    "all_countries_list = price_parser.get_country_list()\n",
    "print(f\"Number of countries: {len(all_countries_list)}\")\n",
    "\n",
    "# Date range\n",
    "time_range = \"2015-01-01,2024-12-31\"\n",
    "price_matrix = price_parser.get_price_matrix(time_range=time_range, countries=all_countries_list, fill_method=\"ffill\")\n",
    "returns = price_matrix.pct_change().dropna()\n",
    "weather_features = ['temperature_2m_mean', 'temperature_2m_min', 'temperature_2m_max','precipitation_mean', 'precipitation_min','precipitation_max', 'wind_speed_mean', 'wind_speed_min', 'wind_speed_max']\n",
    "print(f\"Number of total features: {len(weather_features) + 1}\")\n",
    "\n",
    "weather_matrix = weather_parser.get_weather_matrix(\n",
    "    time_range=time_range,\n",
    "    countries=all_countries_list,\n",
    "    fill_method=\"ffill\",\n",
    "    features=weather_features\n",
    ")\n",
    "\n",
    "# 2. Generate cointegration residuals\n",
    "residual_generator = CointegrationResidualGenerator(price_matrix)\n",
    "residual_generator.compute_all_asset_residuals()\n",
    "asset_residuals = residual_generator.get_asset_residuals() # [num_days, num_assets (= countries)]\n",
    "\n",
    "# 3. Prepare data\n",
    "data_prep = DataPreparation(\n",
    "    price_residuals=asset_residuals,\n",
    "    weather_data=weather_matrix,\n",
    "    countries=all_countries_list,\n",
    "    weather_features=weather_features\n",
    ")\n",
    "window_size = 30\n",
    "stride = 1\n",
    "combined_data = data_prep.prepare_rolling_windows(window_size=window_size, stride=stride)\n",
    "# [num_samples, num_features, window_size]\n",
    "\n",
    "next_day_returns = data_prep.prepare_next_day_returns(returns=returns, window_size=window_size, stride=stride)\n",
    "# [num_samples, num_countries]\n",
    "\n",
    "(train_data, train_returns), (val_data, val_returns), (test_data, test_returns) = \\\n",
    "    data_prep.create_train_val_test_split(combined_data=combined_data, next_day_returns=next_day_returns)\n",
    "\n",
    "# Convert to tensors\n",
    "train_data_tensor = torch.FloatTensor(train_data).to(device)\n",
    "train_returns_tensor = torch.FloatTensor(train_returns).to(device)\n",
    "val_data_tensor = torch.FloatTensor(val_data).to(device)\n",
    "val_returns_tensor = torch.FloatTensor(val_returns).to(device)\n",
    "test_data_tensor = torch.FloatTensor(test_data).to(device)\n",
    "test_returns_tensor = torch.FloatTensor(test_returns).to(device)\n",
    "\n",
    "print(f\"Train data shape: {train_data_tensor.shape}\")\n",
    "print(f\"Train returns shape: {train_returns_tensor.shape}\")\n",
    "\n",
    "# 4. Initialize portfolio optimizer\n",
    "portfolio_optimizer = PortfolioOptimizer(\n",
    "    window_size=window_size,\n",
    "    num_countries=len(all_countries_list),\n",
    "    num_weather_features=len(weather_features),\n",
    "    num_filters=8,  # Default value\n",
    "    filter_size=3,\n",
    "    hidden_dim=64,\n",
    "    num_heads=4,\n",
    "    use_transformer=True,\n",
    "    device=device\n",
    ")\n",
    "\n",
    "# 5. Set up trainer with grid search\n",
    "trainer = Trainer(\n",
    "    optimizer=portfolio_optimizer,\n",
    "    train_data=train_data_tensor,\n",
    "    train_returns=train_returns_tensor,\n",
    "    val_data=val_data_tensor,\n",
    "    val_returns=val_returns_tensor,\n",
    "    lr=0.001,\n",
    "    num_epochs=100,\n",
    "    batch_size=32,\n",
    "    patience=50,\n",
    "    device=device,\n",
    ")\n",
    "\n",
    "# 6. Perform grid search\n",
    "param_grid = {\n",
    "    'num_filters': [8, 16, 32],\n",
    "    'filter_size': [3, 5, 7],\n",
    "    'hidden_dim': [64, 128]\n",
    "}\n",
    "best_params, best_score, best_returns = trainer.grid_search(param_grid, verbose=False)\n",
    "print(f\"Best parameters: {best_params} with Sharpe Ratio: {best_score:.4f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a997a7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 7. Reinitialize with best parameters and train on train+val\n",
    "portfolio_optimizer = PortfolioOptimizer(\n",
    "    window_size=window_size,\n",
    "    num_countries=len(all_countries_list),\n",
    "    num_weather_features=len(weather_features),\n",
    "    num_filters=best_params['num_filters'],\n",
    "    filter_size=best_params['filter_size'],\n",
    "    hidden_dim=best_params['hidden_dim'],\n",
    "    num_heads=4,\n",
    "    use_transformer=True,\n",
    "    device=device\n",
    ")\n",
    "trainer = Trainer(\n",
    "    optimizer=portfolio_optimizer,\n",
    "    train_data=torch.cat([train_data_tensor, val_data_tensor]),  # Combine train and val\n",
    "    train_returns=torch.cat([train_returns_tensor, val_returns_tensor]),\n",
    "    val_data=val_data_tensor,  # Still use val for early stopping\n",
    "    val_returns=val_returns_tensor,\n",
    "    lr=0.001,\n",
    "    num_epochs=100,\n",
    "    batch_size=32,\n",
    "    patience=50,\n",
    "    device=device\n",
    ")\n",
    "final_sharpe, _ = trainer.train(verbose=True)\n",
    "\n",
    "# 8. Evaluate on test set\n",
    "test_sharpe, test_returns = trainer.test(test_data_tensor, test_returns_tensor)\n",
    "print(f\"Final Sharpe Ratio on Test Set: {test_sharpe:.4f}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
